packages/devtools_app/lib/src/debugger/codeview.dart

```diff
@@ -243,81 +244,93 @@ class _CodeViewState extends State<CodeView>
 
     _updateScrollPosition(animate: false);
 
-    return OutlineDecoration(
-      child: Column(
-        children: [
-          buildCodeviewTitle(theme),
-          if (lines.isNotEmpty)
-            DefaultTextStyle(
-              style: theme.fixedFontStyle,
-              child: Expanded(
-                child: Scrollbar(
-                  controller: textController,
-                  child: ValueListenableBuilder<StackFrameAndSourcePosition>(
-                    valueListenable: widget.controller.selectedStackFrame,
-                    builder: (context, frame, _) {
-                      final pausedFrame = frame == null
-                          ? null
-                          : (frame.scriptRef == scriptRef ? frame : null);
-
-                      return Row(
-                        children: [
-                          ValueListenableBuilder<
-                              List<BreakpointAndSourcePosition>>(
-                            valueListenable:
-                                widget.controller.breakpointsWithLocation,
-                            builder: (context, breakpoints, _) {
-                              return Gutter(
-                                gutterWidth: gutterWidth,
-                                scrollController: gutterController,
-                                lineCount: lines.length,
+    return HistoryViewport(
+      history: widget.controller.scriptsHistory,
+      generateTitle: (script) => script.uri,
+      buildControls: (context) {
```

Comment:
instead of passing in buildControls and buildContents as parameters, can we just build these in the HistoryViewport widget itself?

Reply:
`buildContents` needs to be a builder as we need to rebuild the content whenever the currently selected historical data changes. `buildControls` is there since this is supposed to be generalized to be used with any historical data and it wouldn't make sense to have script specific controls for non-script history (e.g., navigation history for VM service objects).

Reply:
`buildControls` could take a `Row` instead of a `Row Function(BuildContext)` though.